python/lsst/daf/butler/_butler.py

```diff
@@ -525,6 +528,83 @@ def get_known_repos(cls) -> set[str]:
         """
         return ButlerRepoIndex.get_known_repos()
 
+    @classmethod
+    def parse_dataset_uri(cls, uri: str) -> tuple[str, DatasetId]:
+        """Extract the butler label and dataset ID from a dataset URI.
+
+        Parameters
+        ----------
+        uri : `str`
+            The dataset URI to parse.
+
+        Returns
+        -------
+        label : `str`
+            The label associated with the butler repository from which this
+            dataset originates.
+        dataset_id : `DatasetId`
+            The ID of the dataset.
+
+        Notes
+        -----
+        Supports dataset URIs of the forms ``ivo://rubin/butler_label/UUID``
+        and ``butler://butler_label/UUID``. In ``ivo`` URIs the butler label
+        can include ``/`` and the leading ``/`` is always stripped. If the
+        repository label starts with ``/`` then it must be doubled up. e.g.,
+
+            ivo://rubin//repo/main/82d79caa-0823-4300-9874-67b737367ee0
```

Comment:
Following discussion with @timj today, the suggestion is that the netloc be treated leniently -- it will have a standard value (e.g., "rubin", still to be confirmed with the IVOA), and then the path will be:

butler/(butler-identity)/(uuid)

If the *butler-identity* begins with a "/", e.g., "/repo/main", then an extra "/" is required in the pathname, e.g., 

ivo://rubin/butler//repo/main/82d79caa-0823-4300-9874-67b737367ee0

To be explicit the *butler-identity* is everything between the first and last slashes in the path.

Reply:
After discussing this with Markus it was noted that the variable part of the IVO ID has to be a query parameter. ie something like

```
ivo://lsst.org/butler?/repo/main/82d79caa-0823-4300-9874-67b737367ee0
```
or I suppose:
```
ivo://lsst.org/butler//repo/main?82d79caa-0823-4300-9874-67b737367ee0
```
Depending on whether we have a look up service per butler concept or per butler repository.
Maybe it's worth pondering what it would look like for an Object/Source ID.
```
ivo://lsst.org/object?1234567890
```
where in theory we can work out the data release number from the object ID. Or
```
ivo://lsst.org/object/dr1?1234567890
```

Reply:
or even

```
ivo://lsst.org/dr1?1234567890
```
and
```
ivo://lsst.org/dr1?butler/dr1/82d79caa-0823-4300-9874-67b737367ee0
```

Reply:
I am working through ideas for IVO IDs on [DMTN-302](https://dmtn-302.lsst.io).

Reply:
For completeness we now parse `ivo://org.rubinobs/usdac/lsst-dr1?repo=dr1id=UUID`.

The parser does not attempt to look at anything other than the query component.